### PR TITLE
Convert to multi-stage builds

### PR DESCRIPTION
**Pull request by lukebakken**
Start with the Ubuntu template

Follow-up to #587

### CODE REVIEW

**Comment by lukebakken** [comment]
@tianon I've got some minimal changes in the templates to support multi-stage builds.

I'm not sure what's up with the CI failure, however:

```
bashbrew version v0.1.5
Processing rabbitmq:3.11.3 ...
failed fetching repo "ubuntu:20.04"
unable to find a manifest named "ubuntu" (in "/tmp/tmp.SIfMOjytit/library" or as a remote URL)
+ strategy=
```

...I can reproduce the same error locally, but I'm not sure of the source of the error. It _seems_ like `generate-stackbrew-library.sh` is working correctly.

Note that I had to add the `setup-go` action to fix a compile error in bashbrew. Apparently go 1.18 is required now? 🤷‍♂️ 

Assistance is appreciated!

**Comment by tianon** [comment]
Sorry!  If you drop your `.github` changes and rebase on https://github.com/docker-library/rabbitmq/pull/590, it should be fixed. :bow:

**Comment by lukebakken** [comment]
Thanks!

I'm curious why the Alpine builds specify `3.16` rather than just `3`? Has there been a problem in the past?

**Comment by lukebakken** [comment]
This is just about ready for review. I need to figure out why the alpine image is quite a bit larger with these changes in my env:

```
$ docker images '*rabbit*'

REPOSITORY                                  TAG                        IMAGE ID       CREATED         SIZE
rabbitmq-local-multi-stage                  alpine-latest              fd57cc784e62   2 minutes ago   162MB
rabbitmq                                    3-alpine                   6d8788289676   4 days ago      127MB

rabbitmq-local-multi-stage                  ubuntu-latest              48b7a0359d68   5 hours ago     229MB
rabbitmq                                    3                          792f1c5d392c   6 days ago      229MB
```

**Comment by lukebakken** [comment]
Within the running containers the sizes are comparable:

```
# rabbitmq:3-alpine    # alpine multi-stage
bash-5.1# du -hs *     bash-5.1# du -hs *
2.0M    bin            2.0M    bin
0       dev            0       dev
547.0K  etc            675.0K  etc
512     home           512     home
4.2M    lib            4.2M    lib
2.0K    media          2.0K    media
512     mnt            512     mnt
24.5M   opt            24.5M   opt
512     plugins        512     plugins
1.0K    root           1.0K    root
512     run            512     run
160.0K  sbin           171.0K  sbin
512     srv            512     srv
0       sys            0       sys
1.0K    tmp            1.0K    tmp
98.8M   usr            98.3M   usr
108.0K  var            108.0K  var
```

**Comment by lukebakken** [comment]
I'm not going to worry too much about the local image size because if I build from `master` in the `3.11/alpine` directory the size is larger than the pulled image 🤷‍♂️

File: Dockerfile-alpine.template
Comment (tianon) [inline Dockerfile-alpine.template:114]: Since we're multi-stage, it doesn't hurt us to add an extra layer here, and what we get in return is the ability to inspect the failed compilation a little:

```suggestion
	ln -sf /etc/ssl/certs /etc/ssl/private "$OPENSSL_CONFIG_DIR"

￼# smoke test
RUN openssl version
```

File: Dockerfile-alpine.template
Comment (tianon) [inline Dockerfile-alpine.template:191]: Similar to the above suggestion :eyes:

```suggestion
	find /usr/local/lib/erlang -type d -name include -exec rm -rf '{}' +

# Check that Erlang/OTP crypto & ssl were compiled against OpenSSL correctly
RUN erl -noshell -eval 'io:format("~p~n~n~p~n~n", [crypto:supports(), ssl:versions()]), init:stop().'
```

File: Dockerfile-alpine.template
Comment (tianon) [inline Dockerfile-alpine.template:211]: Doing `readonly` this way won't interact with `set -e` correctly, so it needs to be after the variable is set:

```console
$ docker run --rm alpine:3.16 sh -c 'set -eux; readonly foo="$(false)"; echo oops'
+ false
+ readonly 'foo='
+ echo oops
oops

$ docker run --rm alpine:3.16 sh -c 'set -eux; foo="$(false)"; readonly foo; echo oops'
+ false
+ foo=

$ docker run --rm alpine:3.16 sh -c 'set -eux; foo="$(echo nice)"; readonly foo; echo "$foo"'
+ echo nice
+ foo=nice
+ readonly foo
+ echo nice
nice
```

```suggestion
	runDeps="$( \
￼		scanelf --needed --nobanner --format '%n#p' --recursive /usr/local \
￼			| tr ',' '\n' \
￼			| sort -u \
￼			| awk 'system("test -e /usr/local/lib/" $1) == 0 { next } { print "so:" $1 }' \
￼	)"; \
	readonly runDeps; \
```

File: Dockerfile-alpine.template
Comment (tianon) [inline Dockerfile-alpine.template:251]: I don't have a problem with these being here, but I think they'd be slightly better right up above `ENV RABBITMQ_VERSION` (for the cache behavior).

File: Dockerfile-ubuntu.template
Comment (tianon) [inline Dockerfile-ubuntu.template:116]: Similar suggestion as Alpine for making this a separate smoke test layer :eyes:

File: Dockerfile-ubuntu.template
Comment (tianon) [inline Dockerfile-ubuntu.template:215]: Don't we still need this list saved somewhere (while we have all the deps installed) so we can install them in the later stage? (I'm guessing the actual list of deps must be already satisfied in the base image as-is? :sweat_smile:)

File: generate-stackbrew-library.sh
Comment (tianon) [inline generate-stackbrew-library.sh:59]: Long-term we might want to rewrite this to use a hash lookup table to track `AS` values so we can ignore them in subsequent `FROM` instructions, but this should be fine for now (technically, we can also have remote images / remote dependencies in `COPY --from=`, but I don't see a strong reason for us to use those here yet so we can cross that entire bridge if/when we come to it). :+1:

**Review by tianon — COMMENTED** [review]
Overall pretty minor comments all around! :+1:

I did compare CI builds of yours to recent master CI builds, and the sizes look comparable:
- https://github.com/docker-library/rabbitmq/actions/runs/3483776129/jobs/5827634135#step:8:19 (your `3-alpine`, ~126MiB)
- https://github.com/docker-library/rabbitmq/actions/runs/3483672593/jobs/5827423409#step:8:19 (master's `3-alpine`, ~128MiB)
- Ubuntu-based builds were similar, yours at ~228MiB, master at ~229MiB

:+1:

**Comment by tianon** [comment]
(To be explicit: I think the `readonly` suggestion is the only one of those I'd make a blocker - everything else is fine as-is if you don't want to change it. :+1:)

File: Dockerfile-alpine.template
Comment (lukebakken) [reply Dockerfile-alpine.template:114]: Great idea

File: Dockerfile-alpine.template
Comment (lukebakken) [reply Dockerfile-alpine.template:211]: Good catch. No reason to use `readonly` tbh

File: Dockerfile-ubuntu.template
Comment (lukebakken) [reply Dockerfile-ubuntu.template:215]: Yep I will restore that! Whoops.

Comment (lukebakken) [reply Dockerfile-ubuntu.template:215]: It's interesting, it does work as-is, but just because we're getting lucky.

**Comment by lukebakken** [comment]
I've always wondered exactly why `shellcheck flags that use of `readonly`, and thanks to your example I understand why now.

**Comment by lukebakken** [comment]
@tianon thanks for the review. Seems like multiple builds are being generated for the same Docker image?

https://github.com/docker-library/rabbitmq/actions/runs/3500314277

...not sure if that's due to my changes or something in `bashbrew`

**Comment by tianon** [comment]
> Seems like multiple builds are being generated for the same Docker image?

That's intentional -- note the `(i386)` at the end.  It's doing a 32bit build for images which support it (in our case only Alpine since Ubuntu dropped it) as a rough smoke test for our multiarch support. :+1:

File: Dockerfile-ubuntu.template
Comment (tianon) [inline Dockerfile-ubuntu.template:214]: Sorry, I should've been more specific -- if we ever stop getting lucky and not having our deps in the base image already, this won't quite work as-is because `ldd` won't list the ones we're missing in the same way it did while the deps are installed, _and_ we can't then `dpkg-query --search` to ask which package they come from without having them installed, so we'd need to do something like gather the list of APT dependencies in the previous stage into a text file or something so we can install it here.

(On the other hand, Alpine doesn't suffer from this because it uses `readelf` to get a list of `.so` dependencies and the package system there includes the `so:...` aliases we take advantage of, so it "just works" appropriately there in ways Ubuntu can't. :sweat_smile:)

I'm totally on board with just dropping this completely from Ubuntu for now (since our `openssl version` and/or `erl` tests below will catch any problems and fail before we publish broken images), and then we revisit the implementation later, or if you wanted to pick an appropriate place to have the previous stage write an `apt-deps.txt` or similar that we can do something like `xargs < /.../apt-deps.txt -rt apt-get install --yes --no-install-recommends` on here.  Your choice! :smile:

Comment (lukebakken) [reply Dockerfile-ubuntu.template:214]: Well that's interesting. Seems like the simplest route is to just nuke it and let the tests catch it.

**Review by tianon — APPROVED** [review]

File: Dockerfile-alpine.template
Comment (yosifkit) [inline Dockerfile-alpine.template:13]: I'm not certain what `openssl-builder` being separate from `build-base` target really adds. It seems unlikely that we'd need to only build `build-base` while debugging. This applies to the Ubuntu Dockerfile as well.

We technically only need just a single target for the compiled artifacts, but it seems that having two would be slightly easier during Dockerfile development (for something like OpenSSL 3 changes).

Comment (lukebakken) [reply Dockerfile-alpine.template:13]: > it seems that having two would be slightly easier during Dockerfile development (for something like OpenSSL 3 changes)

That's the rationale.

**Comment by lukebakken** [comment]
@tianon I'll resolve the conflicts. Anything else need to be done?

**Comment by lukebakken** [comment]
Hmm, investigating the latest round of alpine build failures.

**Comment by tianon** [comment]
Ah, that's probably #592 -- Alpine 3.17 updated the default OpenSSL in the distro to OpenSSL 3, so maybe that's conflicting somehow (I thought we had some mitigations to avoid that but maybe they broke or something)?

**Comment by lukebakken** [comment]
@tianon ha, no it was much simpler... I didn't update the final base image to `alpine:3.17`, so there were mismatches in which packages were added for Erlang run-time deps.

It did prompt me to ask this - https://github.com/docker-library/rabbitmq/discussions/596

**Comment by lukebakken** [comment]
@tianon are we waiting on one more review here? Do you need anything else from me?

**Comment by tianon** [comment]
Yeah, just waiting for @yosifkit to get some time to go through it again :+1:

File: Dockerfile-ubuntu.template
Comment (yosifkit) [inline Dockerfile-ubuntu.template:198]: Just a future note that I'd love to improve this with `--link`, but sadly it only effects the transfer to the build cache and doesn't improve the image like it is purported to do :cry:(https://docs.docker.com/engine/reference/builder/#copy---link).

It might be working in the next? release (https://github.com/moby/buildkit/issues/2884).

**Review by yosifkit — APPROVED** [review]

File: Dockerfile-ubuntu.template
Comment (lukebakken) [reply Dockerfile-ubuntu.template:198]: Yep I had that in my first "let's change a LOT" iteration 😉

**Pull request merged**
